File: frontend/src/app/services/categoria.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Categoria } from '../models/categoria.model';
import { environment } from '../../environments/environment';

@Injectable({ providedIn: 'root' })
export class CategoriaService {
  private apiUrl = `${environment.apiUrl}/categorias`;

  constructor(private http: HttpClient) {}

  listarTodas(): Observable<Categoria[]> {
    return this.http.get<Categoria[]>(this.apiUrl);
  }

  listarPorTipo(tipo: string): Observable<Categoria[]> {
    return this.http.get<Categoria[]>(`${this.apiUrl}/tipo`, { params: { tipo } });
  }

  crear(nombre: string, tipo: string): Observable<Categoria> {
    return this.http.post<Categoria>(this.apiUrl, null, { params: { nombre, tipo } });
  }

  actualizar(id: number, nombre: string, tipo: string): Observable<Categoria> {
    return this.http.put<Categoria>(`${this.apiUrl}/${id}`, null, { params: { nombre, tipo } });
  }

  eliminar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  obtenerPorId(id: number): Observable<Categoria> {
    return this.http.get<Categoria>(`${this.apiUrl}/${id}`);
  }
}

File: src/main/java/com/ucc/finanzas/service/CategoriaService.java
package com.ucc.finanzas.service;

import com.ucc.finanzas.model.Categoria;
import com.ucc.finanzas.repository.CategoriaRepository;
import org.springframework.stereotype.Service;

import java.util.List;

/**
 * Servicio de categorias.
 * Permite crear, editar, eliminar y listar las categorias que clasifican ingresos y gastos.
 */
@Service
public class CategoriaService {

    private final CategoriaRepository categoriaRepository;

    public CategoriaService(CategoriaRepository categoriaRepository) {
        this.categoriaRepository = categoriaRepository;
    }

    public Categoria crear(String nombre, String tipo) {
        Categoria categoria = new Categoria(nombre, tipo.toUpperCase());

        if (!categoria.validarNombre()) {
            throw new RuntimeException("El nombre de la categoria no puede estar vacio.");
        }
        if (!categoria.validarTipo()) {
            throw new RuntimeException("El tipo de categoria solo puede ser INGRESO o GASTO.");
        }

        return categoriaRepository.save(categoria);
    }

    public Categoria actualizar(Long id, String nombre, String tipo) {
        Categoria categoria = categoriaRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Categoria no encontrada con id: " + id));

        categoria.setNombre(nombre);
        categoria.setTipo(tipo.toUpperCase());

        if (!categoria.validarNombre()) {
            throw new RuntimeException("El nombre de la categoria no puede estar vacio.");
        }
        if (!categoria.validarTipo()) {
            throw new RuntimeException("El tipo de categoria solo puede ser INGRESO o GASTO.");
        }

        return categoriaRepository.save(categoria);
    }

    public void eliminar(Long id) {
        if (!categoriaRepository.existsById(id)) {
            throw new RuntimeException("Categoria no encontrada con id: " + id);
        }
        categoriaRepository.deleteById(id);
    }

    public Categoria obtenerPorId(Long id) {
        return categoriaRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Categoria no encontrada con id: " + id));
    }

    public List<Categoria> listarTodas() {
        return categoriaRepository.findAll();
    }

    public List<Categoria> listarPorTipo(String tipo) {
        return categoriaRepository.findByTipo(tipo.toUpperCase());
    }
}
